### PR TITLE
Correctly adjust window size based on the computed visible area for d…

### PR DESCRIPTION
**Pull request by antonioginer**
…ifferent stretching cases.

Fixes https://github.com/mamedev/mame/issues/9115

On Windows, it will also resize the window when keepaspect is changed from the ui.

### CODE REVIEW

**Comment by cd4053** [comment]
Hi, this is what I got.
Command: `mame sf2 -norc -noues -skip_gameinfo -window -sound none -pause_brightness 1.0`
Artwork: [circles.zip](https://github.com/mamedev/mame/files/7859288/circles.zip)

**Issue 1**
If you turn off "Non-Integer Scaling" the aspect ratio is off, the screen area return a non-integer scaling instead of an integer one. From that screen we have a **1152 x 896 (9:7)**, if we take `896 * (4/3) = 1194,66666666667`. For integer we need to do `896 + 1 * (4/3) = 1196`. For **integer scaling** while "Maintain Aspect Ratio On" should have something like **1196 x 897 (4:3)**, or maybe any other 4:3 integer value.
image

For easy reference this is how it looks like with some circles on screen.
Command: `mame sf2 -norc -noues -skip_gameinfo -window -sound none -pause_brightness 1.0 -override_artwork circles`
**Non-Integer Scaling Off** / **Maintain Aspect Ratio On**
image

**Non-Integer Scaling On** / **Maintain Aspect Ratio On**
image

**Issue 2**
With vertical screen we have a similar issue for integer scaling.
image
image
image

**Issue 3**
Rotating the screen breaks aspect ratio for integer scaling.
image
image

**Comment by antonioginer** [comment]
Those are non-issues, all what you're seeing is expected.

> **Issue 1** If you turn off "Non-Integer Scaling" the aspect ratio is off, the screen area return a non-integer scaling instead of an integer one. From that screen we have a **1152 x 896 (9:7)**, if we take `896 * (4/3) = 1194,66666666667`. For integer we need to do `896 + 1 * (4/3) = 1196`. For **integer scaling** while "Maintain Aspect Ratio On" should have something like **1196 x 897 (4:3)**, or maybe any other 4:3 integer value.

That's not the purpose of integer scaling. What integer scaling means is to apply integer **factors** to both axes. In that situation, if you enable `-keepaspect`, the algorithm will try to **approximate** the source aspect ratio, but in most cases it won't get a perfect match.

You probably have a 1920x1080 resolution screen. On that resolution, the best integer match is **3x : 4x,** which results in 1152x896, aspect ratio 1.286 vs 1.333.

The first integer pair that provides a perfect match for sf2 is **7x : 9x**, which results in 2688x2016 (1.333). On a 4K screen, that's exactly what you get with `-noues`.

In short, the code is doing its job.

Let me remind that full integer scaling was implemented to support CRTs, where non-square pixels is the norm. For LCD screens, please just use the other 2 options that apply integer scaling to just one of the axes.

**Comment by cd4053** [comment]
> Those are non-issues, all what you're seeing is expected.

Aspect ratio still broken though.

> That's not the purpose of integer scaling. What integer scaling means is to apply integer factors to both axes. 

So, mame actually does a **integer-ratio scaling**.

> In that situation, if you enable -keepaspect, the algorithm will try to approximate the source aspect ratio, but in most cases it won't get a perfect match.

If the algorithm uses SAR (Source/Storage Aspect Ratio), we could have a 4:3 (or 3:4) DAR (Display Aspect Ratio) correction option because after all, we use DAR. Maybe change "Maintain Aspect Ratio" to "Maintain **SAR** Aspect Ratio" so a regular user or everybody else could understand, precisely, what the option does.

> You probably have a 1920x1080 resolution screen. On that resolution, the best integer match is 3x : 4x, which results in 1152x896, aspect ratio 1.286 vs 1.333.

Actually it's **5x** because `1080 / 224 = 4,8214... `, if we use **5x** we have:

`1920 x 1120` = `SAR:` **12:7**, `PAR:` **1,714** 

If we take the game resolution of SF2 `384 x 224`:

`384 x 224` = `SAR:` **12:7**, `PAR:` **1,714**

As you can see, **x5** maintain a **perfectly** SAR and PAR.

As you demonstrated, the current algorithm returns:

`1152 x 896` = `SAR:` **9:7**, `PAR:` **1,286**

This doesn't maintain the original SAR/PAR (inaccurate), as we can also see in the screenshots.

Even though 5x gives a correct integer-ratio scaling for SAR/PAR, we still need to have an option for DAR correction since games of that era were made for DAR 4:3 or PAR 1,333.

> For LCD screens, please just use the other 2 options that apply integer scaling to just one of the axes.

Still returns a broken AR:

`-noues -sx 3`

image

`-noues -sy 3`

image

If you allow me to give some sugestivos:

- Rename "Non-Integer Scaling" to "Integer-Ratio Scaling".
- Rename "Maintain Aspect Ratio" to "Maintain SAR Aspect Ratio".
- Add a option to allow DAR correction like "Correct/Fix/Maintain/??? DAR Aspect Ratio".
- If DAR option is enabled, disable SAR and vice-versa.

**Thank you** @antonioginer for addressing the issue and for your time.

**Comment by cd4053** [comment]
I'm sorry, I forgot to reply to this one.

> Let me remind that full integer scaling was implemented to support CRTs, where non-square pixels is the norm.

We can have both worlds, create an option to "support CRTs" so it uses an algorithm for people that uses CRT's, this way people with CRT's and LCD's can enjoy correct aspect ratio for both worlds.

Thank you again.

**Comment by antonioginer** [comment]
@cd4053

The aspect ratio isn't broken. Integer scaling isn't arithmetically possible while keeping aspect in most cases, you can't fight maths.

As long as you use `-noues`, you're going to face this issue. That's why I told you _**not**_ to use it, but instead, use the other pair of options:

`-unevestretchx (-uesx)`      (fractional x : integer y) use for horizontal games
`-unevestretchy (-uesy)`      (integer x : fractional y) use for vertical games
`-autostretchxy`    let mame decide between -uesx and -uesy depending on game's orientation.

Those 2 will provide you with the correct aspect ratio, and still apply integer scaling in the axis that matters in each case (which depends on the game's native orientation, i.e. direction of scanlines).

**Comment by antonioginer** [comment]
I agree that the terminology is confusing, because the names of these features are defined in negative (no-features), for historical reasons.

Instead of `-intscale` (the feature), we call it `-nounevenstretch` (the no-feature)
Instead of `-intscaley` (the feature happens on y), we call it `-unevenstretchx` (the no-feature happens on x)
etc.

**Comment by antonioginer** [comment]
> If you allow me to give some sugestivos:
> 
> * Rename "Non-Integer Scaling" to "Integer-Ratio Scaling".
> * Rename "Maintain Aspect Ratio" to "Maintain SAR Aspect Ratio".
> * Add a option to allow DAR correction like "Correct/Fix/Maintain/??? DAR Aspect Ratio".
> * If DAR option is enabled, disable SAR and vice-versa.

I don't fully get your SAR vs DAR distinction.

I usually stay away of discussions about aspect ratio because they inevitably turn into semantics. We may be talking about the same things but from different perspectives.

I believe the DAR correction you suggest is what `-uesx`, `-uesy` along with `-keepaspect` already do.

"Non-Integer Scaling" should be called "Fractional Scaling" (so you'd have fractional vs integer). That sounds the most intuitive to me, though I might be biased by my background.

**Comment by cd4053** [comment]
Hi @antonioginer thank you for answering.

> The aspect ratio isn't broken. Integer scaling isn't arithmetically possible while keeping aspect in most cases, you can't fight maths.

As the images has shown, even though the algorithm is doing what is suppose to, the defaults display a broken AR, the circles artwork give you a cue of how much it get distorted. Maybe that can be fixed by changing mame's default from `-unevenstretch (-ues)` to `-autostretchxy (-asxy)` and let the algorithm decide how to handle it and solve a lot of problems related with that.

> As long as you use -noues, you're going to face this issue. That's why I told you not to use -noues

Since you have address the white elephant in the room, people (me included) uses `-noues` to avoid the problem it causes, excuse me for the lack of technical term, I'm not a software Engineer, it distorts the pixels across the screen. Maybe playing on a CRT or small LCD people will not notice however, you can't ignore it if you use a big screen with no filter on.

Using this as reference:
ken

I took a few pixels from the left of Ken's head to illustrate:
pixels

On the left image we have `-unevenstretch (-ues)`, the midde  `-autostretchxy (-asxy)` and the one the right `-nounevenstretch (-noues)` with no AR corrected.  For accuracy sake, since that distortion is introduced by the algorithm, we disable it. It's better to use `-noues` and have a DAR 4:3 or PAR 1,333 correction instead.

> -unevestretchx (-uesx) (fractional x : integer y) use for horizontal games
-unevestretchy (-uesy) (integer x : fractional y) use for vertical games
-autostretchxy let mame decide between -uesx and -uesy depending on game's orientation.

Thank you for addressing that, now we know exactly how it works.

> I don't fully get your SAR vs DAR distinction.

By analogy, is something similar to [anamorphic format](https://en.wikipedia.org/wiki/Anamorphic_format), you register something distorted so later it will be stretched by a fixed amount. **SAR** is how the image it's generated by the board and **DAR** is how it would be displayed on a 4:3 monitor. This is why SAR and DAR distinction, because games of that time were designed to de displayed (DAR) on a 4:3 CRT monitor. More explanation about it [here](https://www.youtube.com/watch?v=LHfPA4n0TRo), [bonus](https://vgdensetsu.tumblr.com/post/178414835138/from-chara-design-to-graphic-design-the-artists).

> I usually stay away of discussions about aspect ratio because they inevitably turn into semantics. We may be talking about the same things but from different perspectives.

For discussion sake, I don't like it either. However, they are not always just perspectives and semantics as we can see here:

`-asxy`
snes

[They suggest something else](https://wiki.nesdev.org/w/index.php/Overscan#For_emulator_developers) related to that ([source](https://www.resetera.com/threads/playing-old-games-on-a-modern-display-black-bars-and-og-aspect-ratio-or-stretch-to-16-9.95241/page-2)):
bLSZUjb

Personally I disagree, CRT TV's at that time were 4:3 AR, if the developer at Nintendo got it wrong, not my problem. As you can see, we can open a can of worms just by talking about it.

> I believe the DAR correction you suggest is what -uesx, -uesy along with -keepaspect already do.

Yes, that works for SAR correction as you have explained. For DAR correction however, it would still display a wrong AR.

> "Non-Integer Scaling" should be called "Fractional Scaling" (so you'd have fractional vs integer). That sounds the most intuitive to me, though I might be biased by my background.

Please do!
I'm all for it.

Thank you.

**Comment by rb6502** [comment]
The longer this discussion goes on the more I'm on team "remove all of these options".

**Comment by cd4053** [comment]
> The longer this discussion goes on the more I'm on team "remove all of these options".

Personally, I already disable the option since day one. I agree, if the option causes all that trouble, remove it.

**Comment by cd4053** [comment]
Hi @ICEknigh7 
**left** `-ues -filter -prescale 8`, **right** `-ues -filter -prescale 4` (mame default):
pixels3

**Comment by cd4053** [comment]
I've found [this article](https://tanalin.com/en/articles/integer-scaling) that describes the exact same issue we are seeing here, [this](https://tanalin.com/en/articles/integer-scaling/#h-partial-emulators-mame) is the part related with mame:

    The emulator of multiple game systems and old computers MAME supports integer-ratio scaling with no blur.
    Integer-ratio scaling can be enabled with the command-line parameter -nounevenstretch or its short version -noues.

Again, we use `-noues` to remove all the problems it causes with pixels.

    The algorithm of integer-ratio scaling in MAME uses pixels of the same size, so at low scales, aspect ratio of the scaled
    image may be inaccurate if the game needs aspect-ratio correction. For example, in the arcade version of the
    “Mortal Kombat 3” game (the original resolution is 400×254, target aspect ratio is 4:3) the scaled image is
    noticeably too narrow horizontally on FHD displays (logical-pixel size is 3×4 physical pixels, an error is 11.42%),
    while such disproportion is unnoticeable on 4K displays (pixel size is 7×8, an error is 3.34%).

The autor identify a similar issue I demonstrated with Street Fighter II, **"_the scaled image is noticeably too narrow horizontally_"**:
image

There is no need to trash the hard work made by our friend @antonioginer, however, **we need a DAR correction algorithm** to address this issue for LCD screens. This way the CRT people can have the "Non-Integer Scaling" or "Fractional Scaling"  option and people that doesn't use it, can turn the option off and have a "Maintain DAR" option to correct the aspect ratio for DAR 4:3 / PAR 1,333 (3:4 depending on the orientation). This way whatever distortion `-noues` causes to the screen, the DAR algorithm fixes.

By default, I think "Non-Integer Scaling" or "Fractional Scaling" should be **off** and "Maintain DAR" should be on so people can have perfect pixels with correct aspect ratio, what do you guys think?

Also, since now we know "Maintain Aspect Ratio" and `-keepaspect` are related with **SAR** ("Maintain SAR"), make sure it got disabled if "Maintain DAR" is enabled and vice-versa.

This way people that want pixel perfect with no distortion and people from CRT can coexist without issues.

Thank you.

**Comment by antonioginer** [comment]
Hey cd4053,

> By default, I think "Non-Integer Scaling" or "Fractional Scaling" should be **off** and "Maintain DAR" should be on so people can have perfect pixels with correct aspect ratio, what do you guys think?

If after my explanations above, it hasn't become crystal clear to you that your statement above is mathematically imposible, I don't know what else I can do.

I believe this is far from your intention, but your perseverance on making these features appear as controversial is going to end up causing the REMOVAL of such features, by some devs that couldn't care less about them.

There is a PR on top of this nonsensical thread, I'd appreciate any subsequent comments have to do with the code in there. Discussion is welcome in other places, not here, please.

**Comment by ICEknigh7** [comment]
Yeah, everything in this PR seems to be working fine for everybody else.

**Comment by cd4053** [comment]
Hi @antonioginer, thank you for answering.

> If after my explanations above, it hasn't become crystal clear to you that your statement above **_is mathematically imposible_**, I don't know what else I can do.

If the algorithm doesn't allow it for some reason (it would break something else in the code), then there is nothing we can do then, keep the broken AR

> I believe this is far from your intention, but your perseverance on making these features appear as controversial is going to end up causing the REMOVAL of such features, by some devs that couldn't care less about them.

Your PR fixes the window issue, thank you for that! However it still displays a **broken AR**, even with the window issue fixed, the display show distortion, the pixels get distorted, jittered, it create distortions _that doesn't exist in the real arcade/console/personal computer_ and in the end it benefits only a small group of people _just for looks_, to make it look good on a CRT display. If the original arcade or console **doesn't** generate that type of pixel distortion, breaking the output aspect ratio as it does now and is impossible to fix, for _accuracy and preservation sake_, this code should no be here.

I'm sorry, I really am. I'm with the devs, I vote and agree with a **complete REMOVAL of this code**, it does more harm than good (as demonstrated).

> There is a PR on top of this nonsensical thread, I'd appreciate any subsequent comments have to do with the code in there. Discussion is welcome in other places, not here, please.

My intention was just to report the issue with broken AR in your algorithm, hoping for a fix and not to cause any troubles.
No more discussion, excuse me if I cause any trouble to you or anybody else, it was not my intention.

Thank you for your time.

**Comment by ICEknigh7** [comment]
@cd4053 You're better off opening an issue in Mametesters with all that stuff, since those things come from before the changes in this PR and thus are out of place.

**Comment by cd4053** [comment]
Hi @ICEknigh7 

> since those things come from before the changes in this PR and thus are out of place.

If you take a moment to read the PR title _"Correctly adjust window size based on the **computed visible area**"_, the computed visible area display are displaying a broken AR when you turn off `-ues`.

> You're better off opening an issue in Mametesters with all that stuff

Sorry, I can't.
I'm just an advanced mame user mother of 3 children doing a report. I don't know how to use Mametesters, how to do a report, how to identify the driver, device or source code to point to, I don't understand code, I'm not a developer or software engineer. The only reason I can talk about aspect ratio is because I'm a designer and we had to deal with this since the 80's converting footage from cinema to TV.

This is what I know, "hey your code is doing this, is causing that", I can't engage in a more technical discussion.

If you or somebody else has the knowledge to publish this on mametesters, please do.

Thank you.

**Comment by Robbbert** [comment]
This argument is unproductive, and should be taken to a forum or some other discussion place. Locked.

**Comment by rb6502** [comment]
The aspect ratio cannot be perfect on an LCD when you're doing pure integer scaling.  This has always been known.  Math doesn't lie.
